### ISSUE TITLE
Retry policy for ShopifyRateLimitException

### ISSUE THREAD

**Issue by clement911** [nozzlegear/ShopifySharp#91]
I understand that currently (v3), it is up to applications to catch potential ShopifyRateLimitException and handle appropriately.
We could augment the current library so that the handling of the rate limit can be configured once centrally, as opposed to the application having to wrap every call to the shopify sharp library...

I was thinking that, at startup, applications might be able to call a new method to configure the policy:
`RequestEngine.SetRateLimitPolicy(myPolicy);`
Then `ExecuteRequestAsync` could use that policy delegate to decide whether to bubble the exception (current behaviour), retry (immediately or after a delay based on the value of the X-Shopify-Shop-Api-Call-Limit header) or maybe something else...

It's quite a high level description here, but I wanted to make sure you were on board before implementing a solution and sending a PR.
What do you think?

**Comment by nozzlegear** [comment]
Hey @clement911, this sounds good to me! Ideally I'd like the default to be 0 retries like it is right now, and then have an option to both set a global retry policy, and an instance-specific retry policy that overrides the global one:

```cs
// Global retry policy
RequestEngine.SetAutomaticRetryLimit(3);

// Instance-specific retry policy
var service = new ShopifyOrderService(domain, token);
service.SetAutomaticRetryLimit(5);
```

If you want to take a crack at this, I'd be happy to accept a PR! If not, I'll definitely get to work on it myself after the holidays.

**Comment by clement911** [comment]
Ok cool I'll give it a go when I get a chance...
In order to be able to override at the service level,  it may be a simpler design if the RequestEngine was not static. This way, each service could have their own RequestEngine instance with their separate retry policy..

**Comment by nozzlegear** [comment]
Good point, I forgot that the RequestEngine is static, so it won't be able to be configured per service right now. I'm not sure if changing the RequestEngine to a non-static class would break somebody's build though. If so, the instance-specific retry policy would have to wait until a 4.0 release.

**Comment by clement911** [comment]
Ok. Should I create a branch 4.0 from master to get started?

**Comment by nozzlegear** [comment]
I've got a few other things to include in 4.0, so for now it'd be best to just implement the policy on the static request engine as it is and I'll make sure instance-specific policy makes it into 4.0.

**Comment by clement911** [comment]
Pull request is here: https://github.com/nozzlegear/ShopifySharp/pull/94

**Comment by nozzlegear** [comment]
Implemented in version 3.5.0 👍
